File: unified-workspace-sdk/tests/e2e-python/test_sandbox.py
#!/usr/bin/env python3
"""E2E tests for Workspace SDK using Python standard library only"""

import os
import sys
import time
import json
import urllib.request
import urllib.error
from typing import Optional, List, Dict, Any
from dataclasses import dataclass
from contextlib import contextmanager

BASE_URL = os.environ.get("WORKSPACE_TEST_URL", "http://127.0.0.1:8080")
BASE_IMAGE = os.environ.get("WORKSPACE_BASE_IMAGE", "docker.easyops.local/ci/rust-builder:1.85.0-centos7")
API_URL = f"{BASE_URL}/api/v1"

@dataclass
class TestResult:
    name: str
    passed: bool
    error: Optional[str] = None
    duration: float = 0.0

results: List[TestResult] = []

@contextmanager
def test_case(name: str):
    """Context manager for running a test case"""
    start = time.time()
    try:
        yield
        duration = time.time() - start
        results.append(TestResult(name=name, passed=True, duration=duration))
        print(f"  ✓ {name} ({duration*1000:.0f}ms)")
    except AssertionError as e:
        duration = time.time() - start
        results.append(TestResult(name=name, passed=False, error=str(e), duration=duration))
        print(f"  ✗ {name} ({duration*1000:.0f}ms)")
        print(f"    Error: {e}")
    except Exception as e:
        duration = time.time() - start
        results.append(TestResult(name=name, passed=False, error=str(e), duration=duration))
        print(f"  ✗ {name} ({duration*1000:.0f}ms)")
        print(f"    Error: {e}")


class WorkspaceClient:
    """Simple HTTP client for Workspace API using standard library"""

    def __init__(self, base_url: str, timeout: float = 60.0):
        self.base_url = base_url
        self.api_url = f"{base_url}/api/v1"
        self.timeout = timeout

    def _request(self, method: str, path: str, data: Optional[Dict] = None) -> Dict[str, Any]:
        url = f"{self.api_url}{path}"
        headers = {"Content-Type": "application/json"}

        if data:
            body = json.dumps(data).encode("utf-8")
        else:
            body = None

        req = urllib.request.Request(url, data=body, headers=headers, method=method)

        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as response:
                return json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            # Re-raise with status code info
            raise HttpError(e.code, e.read().decode("utf-8")) from e

    def health(self) -> Dict[str, Any]:
        return self._request("GET", "/health")

    def create_sandbox(
        self,
        template: Optional[str] = None,
        name: Optional[str] = None,
        env: Optional[Dict[str, str]] = None,
        metadata: Optional[Dict[str, str]] = None,
        timeout: Optional[int] = None,
    ) -> Dict[str, Any]:
        data = {}
        if template:
            data["template"] = template
        if name:
            data["name"] = name
        if env:
            data["env"] = env
        if metadata:
            data["metadata"] = metadata
        if timeout:
            data["timeout"] = timeout

        return self._request("POST", "/sandboxes", data)

    def get_sandbox(self, sandbox_id: str) -> Dict[str, Any]:
        return self._request("GET", f"/sandboxes/{sandbox_id}")

    def list_sandboxes(self) -> Dict[str, Any]:
        return self._request("GET", "/sandboxes")

    def delete_sandbox(self, sandbox_id: str, force: bool = False) -> Dict[str, Any]:
        path = f"/sandboxes/{sandbox_id}"
        if force:
            path += "?force=true"
        return self._request("DELETE", path)


class HttpError(Exception):
    """HTTP error with status code"""
    def __init__(self, status_code: int, message: str):
        self.status_code = status_code
        self.message = message
        super().__init__(f"HTTP {status_code}: {message}")


def main():
    print("Workspace SDK E2E Tests (Python)")
    print("================================")
    print(f"Server URL: {BASE_URL}")
    print(f"Base Image: {BASE_IMAGE}\n")

    client = WorkspaceClient(BASE_URL)
    test_sandbox_id: Optional[str] = None

    try:
        # Health tests
        print("Health Tests:")

        with test_case("health check returns healthy status"):
            health = client.health()
            assert health.get("status") == "healthy", f"Expected 'healthy', got '{health.get('status')}'"
            assert "version" in health, "Version missing from health response"

        # Sandbox tests
        print("\nSandbox Tests:")

        with test_case("create sandbox with name"):
            sandbox = client.create_sandbox(
                template=BASE_IMAGE,
                name="e2e-python-test-sandbox",
            )
            assert sandbox.get("id"), "Sandbox ID is missing"
            assert sandbox.get("name") == "e2e-python-test-sandbox", f"Name mismatch: {sandbox.get('name')}"
            test_sandbox_id = sandbox["id"]

        with test_case("get sandbox by ID"):
            assert test_sandbox_id, "No sandbox ID available"
            sandbox = client.get_sandbox(test_sandbox_id)
            assert sandbox.get("id") == test_sandbox_id, "Sandbox ID mismatch"

        with test_case("list sandboxes includes created sandbox"):
            result = client.list_sandboxes()
            sandboxes = result.get("sandboxes", [])
            found = any(s.get("id") == test_sandbox_id for s in sandboxes)
            assert found, "Created sandbox not found in list"

        with test_case("create sandbox with environment variables"):
            sandbox = client.create_sandbox(
                template=BASE_IMAGE,
                name="e2e-python-env-test",
                env={"MY_VAR": "test_value", "ANOTHER_VAR": "another_value"},
            )
            env = sandbox.get("env", {})
            assert env.get("MY_VAR") == "test_value", f"ENV mismatch: {env}"
            # Cleanup
            client.delete_sandbox(sandbox["id"], force=True)

        with test_case("create sandbox with metadata"):
            sandbox = client.create_sandbox(
                template=BASE_IMAGE,
                name="e2e-python-metadata-test",
                metadata={"project": "test-project", "owner": "test-user"},
            )
            metadata = sandbox.get("metadata", {})
            assert metadata.get("project") == "test-project", f"Metadata mismatch: {metadata}"
            # Cleanup
            client.delete_sandbox(sandbox["id"], force=True)

        with test_case("delete sandbox"):
            assert test_sandbox_id, "No sandbox ID available"
            client.delete_sandbox(test_sandbox_id, force=True)

            # Verify deletion
            try:
                client.get_sandbox(test_sandbox_id)
                assert False, "Sandbox should have been deleted"
            except HttpError as e:
                assert e.status_code == 404, f"Expected 404, got {e.status_code}"
            test_sandbox_id = None

        # Error handling tests
        print("\nError Handling Tests:")

        with test_case("get non-existent sandbox returns 404"):
            try:
                client.get_sandbox("non-existent-sandbox-id")
                assert False, "Expected 404 error"
            except HttpError as e:
                assert e.status_code == 404, f"Expected 404, got {e.status_code}"

    finally:
        # Cleanup any remaining test sandbox
        if test_sandbox_id:
            try:
                client.delete_sandbox(test_sandbox_id, force=True)
            except Exception:
                pass

    # Print summary
    print("\n================================")
    passed = sum(1 for r in results if r.passed)
    failed = sum(1 for r in results if not r.passed)
    total_duration = sum(r.duration for r in results)

    print(f"Results: {passed} passed, {failed} failed")
    print(f"Total time: {total_duration*1000:.0f}ms")

    if failed > 0:
        print("\nFailed tests:")
        for r in results:
            if not r.passed:
                print(f"  - {r.name}: {r.error}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: tests/e2e/test-sandbox.ts
import { WorkspaceClient } from 'workspace-sdk';

const BASE_URL = process.env.WORKSPACE_TEST_URL || 'http://127.0.0.1:8080';
const BASE_IMAGE = process.env.WORKSPACE_BASE_IMAGE || 'rust:1.85';

interface TestResult {
  name: string;
  passed: boolean;
  error?: string;
  duration: number;
}

const results: TestResult[] = [];

async function runTest(name: string, fn: () => Promise<void>): Promise<void> {
  const start = Date.now();
  try {
    await fn();
    results.push({ name, passed: true, duration: Date.now() - start });
    console.log(`  ✓ ${name} (${Date.now() - start}ms)`);
  } catch (error) {
    const errorMsg = error instanceof Error ? error.message : String(error);
    results.push({ name, passed: false, error: errorMsg, duration: Date.now() - start });
    console.log(`  ✗ ${name} (${Date.now() - start}ms)`);
    console.log(`    Error: ${errorMsg}`);
  }
}

async function main(): Promise<void> {
  console.log('Workspace SDK E2E Tests');
  console.log('=======================');
  console.log(`Server URL: ${BASE_URL}`);
  console.log(`Base Image: ${BASE_IMAGE}\n`);

  const client = new WorkspaceClient({ apiUrl: BASE_URL, timeout: 60000 });

  // Health check tests
  console.log('Health Tests:');
  await runTest('health check returns healthy status', async () => {
    const response = await fetch(`${BASE_URL}/api/v1/health`);
    if (!response.ok) {
      throw new Error(`Health check failed: ${response.status}`);
    }
    const data = await response.json() as { status: string };
    if (data.status !== 'healthy') {
      throw new Error(`Expected status "healthy", got "${data.status}"`);
    }
  });

  // Sandbox tests
  console.log('\nSandbox Tests:');
  let testSandboxId: string | null = null;

  await runTest('create sandbox with name', async () => {
    const sandbox = await client.sandbox.create({
      template: BASE_IMAGE,
      name: 'e2e-test-sandbox',
    });
    if (!sandbox.id) {
      throw new Error('Sandbox ID is missing');
    }
    if (sandbox.name !== 'e2e-test-sandbox') {
      throw new Error(`Expected name "e2e-test-sandbox", got "${sandbox.name}"`);
    }
    testSandboxId = sandbox.id;
  });

  await runTest('get sandbox by ID', async () => {
    if (!testSandboxId) throw new Error('No sandbox ID available');
    const sandbox = await client.sandbox.get(testSandboxId);
    if (sandbox.id !== testSandboxId) {
      throw new Error('Sandbox ID mismatch');
    }
  });

  await runTest('list sandboxes includes created sandbox', async () => {
    const { sandboxes } = await client.sandbox.list();
    const found = sandboxes.find(s => s.id === testSandboxId);
    if (!found) {
      throw new Error('Created sandbox not found in list');
    }
  });

  await runTest('create sandbox with environment variables', async () => {
    const sandbox = await client.sandbox.create({
      template: BASE_IMAGE,
      name: 'e2e-test-env',
      env: {
        MY_VAR: 'test_value',
        ANOTHER_VAR: 'another_value',
      },
    });
    if (!sandbox.env || sandbox.env.MY_VAR !== 'test_value') {
      throw new Error('Environment variable not set correctly');
    }
    // Cleanup
    await client.sandbox.delete(sandbox.id, true);
  });

  await runTest('create sandbox with metadata', async () => {
    const sandbox = await client.sandbox.create({
      template: BASE_IMAGE,
      name: 'e2e-test-metadata',
      metadata: {
        project: 'test-project',
        owner: 'test-user',
      },
    });
    if (!sandbox.metadata || sandbox.metadata.project !== 'test-project') {
      throw new Error('Metadata not set correctly');
    }
    // Cleanup
    await client.sandbox.delete(sandbox.id, true);
  });

  await runTest('delete sandbox', async () => {
    if (!testSandboxId) throw new Error('No sandbox ID available');
    await client.sandbox.delete(testSandboxId, true);

    // Verify deletion
    try {
      await client.sandbox.get(testSandboxId);
      throw new Error('Sandbox should have been deleted');
    } catch (error) {
      // Expected: sandbox not found
    }
    testSandboxId = null;
  });

  // Error handling tests
  console.log('\nError Handling Tests:');
  await runTest('get non-existent sandbox returns 404', async () => {
    try {
      await client.sandbox.get('non-existent-sandbox-id');
      throw new Error('Expected error for non-existent sandbox');
    } catch (error) {
      // Expected behavior
      if (error instanceof Error && error.message.includes('Expected error')) {
        throw error;
      }
    }
  });

  // Print summary
  console.log('\n=======================');
  const passed = results.filter(r => r.passed).length;
  const failed = results.filter(r => !r.passed).length;
  const totalDuration = results.reduce((sum, r) => sum + r.duration, 0);

  console.log(`Results: ${passed} passed, ${failed} failed`);
  console.log(`Total time: ${totalDuration}ms`);

  if (failed > 0) {
    console.log('\nFailed tests:');
    results.filter(r => !r.passed).forEach(r => {
      console.log(`  - ${r.name}: ${r.error}`);
    });
    process.exit(1);
  }
}

main().catch(error => {
  console.error('E2E tests failed:', error);
  process.exit(1);
});
